brave/brave-ios
# Fix #6425: Unknown token in ERC20 approve transaction confirmation

## Summary of Changes
- It's possible we have a transaction with a contract address for a token and we don't store the token in our user assets or our blockchain registry. For ERC20 approve, this would cause Transaction Parser to fail to parse and display the empty state for Transaction Confirmation.
- This change makes the tokens optional in parsed transaction models, and if unknown we use the asset ratio service to fetch that token by it's contract address

This pull request fixes #6425

## Submitter Checklist:


## Test Plan:
1. Setup an ERC20 Approve transaction for an unknown token.
    - I used [Uniswap](https://app.uniswap.org/) on Goerli to swap `ETH` -> `UNI`, then when trying to swap `UNI` to `ETH` I need to approve the UNI token (not in our token registry) to perform the swap. Note: Uniswap is first sending a sign message request for the token approval, if you cancel this you should receive a ERC20 Approve transaction after you've cancelled.
2. Tap swap to create the pending transaction
3. Verify that `UNI` token loads in and displays `UNI`
4. Approve, reject or temporarily ignore the transaction. Open Transactions list from the Wallet Panel
5. Verify `UNI` token is shown in the summary, and tap the transaction to open Transaction Detail
6. Verify `UNI` token is shown in the transaction summary


## Screenshots:

https://user-images.githubusercontent.com/5314553/208120099-4af47c8d-fc99-49c5-a075-16d9fafbd105.mp4


## Reviewer Checklist:

  - `QA/(Yes|No)`
  - `bug` / `enhancement`
